### PR TITLE
test(otel): add linking matchers for unit tests

### PR DESCRIPTION
**Pull request by alevenberg**
Part of https://github.com/googleapis/google-cloud-cpp/issues/12528

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12858)

### CODE REVIEW

File: google/cloud/testing_util/opentelemetry_matchers.h [hunk 1]
```diff
@@ -69,6 +69,12 @@ MATCHER_P(EventAttributesImpl, matcher,
                                        result_listener);
 }
 
+MATCHER_P(LinkAttributesImpl, matcher,
```
Comment (dbolduc) [inline google/cloud/testing_util/opentelemetry_matchers.h:72]: No action needed....

These AttributeImpl matchers can be combined if you know how to use templates. Sadly, I cannot help.

File: google/cloud/testing_util/opentelemetry_matchers.h [hunk 2]
```diff
@@ -186,6 +194,38 @@ ::testing::Matcher<SpanDataPtr> SpanEventsAre(Args const&... matchers) {
   return SpanEventsAreImpl(::testing::ElementsAre(matchers...));
 }
 
+MATCHER_P(LinkHasSpanContext, context,
```
Comment (dbolduc) [inline google/cloud/testing_util/opentelemetry_matchers.h:197]: I will wait to see what the tests look like.

If we only ever `LinkHasSpanContext(span->GetContext())`.... then maybe the matcher should be `LinksToSpan(span)`.

**Review by dbolduc — APPROVED** [review]

**Comment by dbolduc** [comment]
Also, FYI you fixed #11236

File: google/cloud/testing_util/opentelemetry_matchers.h [hunk 3]
```diff
@@ -186,6 +194,38 @@ ::testing::Matcher<SpanDataPtr> SpanEventsAre(Args const&... matchers) {
...
+          "has context" + google::cloud::testing_util::ToString(context)) {
+  auto const& actual = arg.GetSpanContext();
+  *result_listener << "has context: "
+                   << google::cloud::testing_util::ToString(actual);
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/testing_util/opentelemetry_matchers.h:201]: Done

https://github.com/googleapis/google-cloud-cpp/pull/12863

Got hit by automerge, whoops.

File: google/cloud/testing_util/opentelemetry_matchers.cc [hunk 4]
```diff
@@ -87,17 +87,28 @@ namespace sdk {
 namespace trace {
 
 std::ostream& operator<<(std::ostream& os, SpanData const& rhs) {
+  char const* line_sep = "\n\t\t\t";
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/testing_util/opentelemetry_matchers.cc:90]: I thought it looked nice 🤷 

Ex:
```
  Actual: { (ptr = 0x55fe0c7e58b0, value = Span {name=Finish, kind=CLIENT, instrumentation_scope {gcloud-cpp, v2.17.0-rc},
			attributes=[gl-cpp.status_code=std::int32_t:0],
			events=[],
			links=[]}), (ptr = 0x55fe0c7e4610, value = Span {name=span, kind=CLIENT, instrumentation_scope {gcloud-cpp, v2.17.0-rc},
			attributes=[gl-cpp.status_code=std::int32_t:10, grpc.peer=std::string:],
			events=[ Event {name=gl-cpp.first-read, attributes=[]}, 
			Event {name=message, attributes=[message.id=std::int32_t:1, message.type=std::string:RECEIVED]}, 
			Event {name=message, attributes=[message.id=std::int32_t:2, message.type=std::string:RECEIVED]}, 
			Event {name=message, attributes=[message.id=std::int32_t:3, message.type=std::string:RECEIVED]}],
			links=[]}) }, where the following matchers don't match any elements:
```

File: google/cloud/testing_util/opentelemetry_matchers.cc [hunk 5]
```diff
@@ -87,17 +87,28 @@ namespace sdk {
...
   for (auto const& e : rhs.GetEvents()) {
     os << sep << "Event {name=" << e.GetName() << ", attributes=["
        << absl::StrJoin(e.GetAttributes(), ", ", AttributeFormatter) << "]}";
-    sep = ", ";
+    sep = ", \n\t\t\t";
+  }
+  os << "]," << line_sep << "links=[";
+  for (auto const& link : rhs.GetLinks()) {
+    os << sep << "Link {span_context="
+       << google::cloud::testing_util::ToString(link.GetSpanContext()) << ","
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/testing_util/opentelemetry_matchers.cc:107]: ditto to my response above.

File: google/cloud/testing_util/opentelemetry_matchers.h (hunk 1 above)
Comment (alevenberg) [reply google/cloud/testing_util/opentelemetry_matchers.h:72]: Yea.... I don't understand it either....

**Pull request merged**